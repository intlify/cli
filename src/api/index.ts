export * from './utils'
export * from './compile'
export * from './annotate'
export * from './format'
