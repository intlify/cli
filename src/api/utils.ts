import type { CompilerError } from '@vue/compiler-sfc'

/**
 * Vue SFC compiler error
 *
 * @remarks
 * This is the error wrapping the error that occurred in Vue SFC compiler
 *
 * @public
 */
export interface SFCParseError extends SyntaxError {
  /**
   * The error that occurred in Vue SFC compiler
   */
  erorrs: CompilerError[]
  /**
   * The filepath of the source file
   */
  filepath: string
}

/**
 * Check wheather Vue SFC compiler error
 *
 * @param err - The error that occurred in Vue SFC compiler
 * @returns if the error is raised in Vue SFC compiler, return `true`, else `false`
 *
 * @public
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isSFCParserError(err: any): err is SFCParseError {
  return 'errors' in err && 'filepath' in err
}
