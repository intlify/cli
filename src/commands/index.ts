export { default as compile } from './compile'
export { default as annotate } from './annotate'
export { default as format } from './format'
