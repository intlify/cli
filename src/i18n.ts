import path from 'path'
import { promises as fs } from 'fs'
import { isString } from '@intlify/shared'
import { createCoreContext, translate } from '@intlify/core'
import { debug as Debug } from 'debug'
import i18nResourceSchema from '../locales/en.json'

import type {
  Locale,
  LocaleMessages,
  LocaleMessageDictionary,
  TranslateOptions,
  CoreContext
} from '@intlify/core'

const debug = Debug('@intlify/cli:i18n')
const DEFAULT_LOCALE = 'en-US'

let resources: LocaleMessages | null = null
let context: CoreContext<unknown, unknown, unknown, string> | null = null

async function loadI18nResources(): Promise<LocaleMessages> {
  const dirents = await fs.readdir(path.resolve(__dirname, '../../locales'), {
    withFileTypes: true
  })
  return dirents.reduce(async (acc, dir) => {
    if (dir.isFile()) {
      const data = await fs.readFile(
        path.resolve(__dirname, '../../locales', dir.name),
        { encoding: 'utf-8' }
      )
      const { name } = path.parse(dir.name)
      debug('load i18n resource', name, data)
      const messages = await acc
      messages[name] = JSON.parse(data) as LocaleMessageDictionary
    }
    return acc
  }, Promise.resolve({} as LocaleMessages))
}

export function getLocale(env?: Record<string, unknown>): Locale {
  env = env || process.env
  const raw =
    ((env.LC_ALL || env.LC_MESSAGES || env.LANG || env.LANGUAGE) as string) ||
    DEFAULT_LOCALE
  debug('getLocale', raw)
  return raw.replace(/\.UTF\-8/g, '').replace(/_/g, '-')
}

export function t<Key extends keyof typeof i18nResourceSchema>(
  key: Key,
  options?: TranslateOptions
): string {
  if (context == null) {
    console.error(
      'cannot initialize CoreContext with @intlify/core createCoreContext'
    )
    return key
  }
  const ret = translate(context, key, options)
  return isString(ret) ? ret : key
}

export async function initI18n() {
  try {
    resources = await loadI18nResources()
    context = createCoreContext({
      locale: getLocale(),
      fallbackLocale: DEFAULT_LOCALE,
      fallbackWarn: false,
      missingWarn: false,
      warnHtmlMessage: false,
      fallbackFormat: true,
      messages: resources
    })
  } catch (e) {
    debug('load i18n resource errors', e.message)
    throw e
  }
}
