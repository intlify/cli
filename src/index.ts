export * from './generator'
export { CompileErrorCodes, CompileOptions, compile } from './api'
