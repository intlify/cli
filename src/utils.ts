import createDebug from 'debug'
import fg from 'fast-glob'
import diff from 'diff-match-patch'
import { parseJSON } from 'jsonc-eslint-parser'
import { parseYAML } from 'yaml-eslint-parser'
import { readFileSync } from 'fs'
import { cosmiconfig } from 'cosmiconfig'
import { promises as fs } from 'fs'
import path from 'pathe'
import { parse } from '@vue/compiler-sfc'
import { isString } from '@intlify/shared'

import type { SFCBlock, SFCDescriptor } from '@vue/compiler-sfc'

declare module '@vue/compiler-sfc' {
  // normalize for `vue-template-compiler`
  interface SFCBlock {
    start?: number
    end?: number
  }
}

const __dirname = path.dirname(new URL(import.meta.url).pathname)
const debug = createDebug('@intlify/cli:utils')

export type CustomBlockContentType = 'json' | 'json5' | 'yaml' | 'unknwon'
export type SFCBloackType =
  | 'template'
  | 'script'
  | 'scriptSetup'
  | 'style'
  | 'custom'
export type SFCAnnotateOptions = {
  type?: string
  force?: boolean
  attrs?: Record<string, any> // eslint-disable-line @typescript-eslint/no-explicit-any
  onWarn?: (msg: string, block: SFCBlock) => void
}
export type parseSFC = (source: string) => ReturnType<typeof parse>

export async function exists(path: string, isThrow = false): Promise<boolean> {
  let ret = false
  try {
    const stat = await fs.stat(path)
    ret = stat.isFile()
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (e: any) {
    if (e.code === 'ENOENT') {
      if (isThrow) {
        throw e
      }
    } else {
      throw e
    }
  }
  return ret
}

type FilterCallback = (file: string | number) => boolean

export async function getSourceFiles(
  input: {
    source?: string
    files?: (string | number)[]
  },
  filter?: FilterCallback | FilterCallback[]
): Promise<string[]> {
  const { source, files } = input
  const _files =
    source != null
      ? await globAsync(source)
      : [...(files || [])].map(a => a.toString()).splice(1)
  if (filter) {
    const _filters = Array.isArray(filter) ? filter : [filter]
    return _filters.reduce((files, filter) => files.filter(filter), _files)
  } else {
    return _files
  }
}

export function globAsync(pattern: string): Promise<string[]> {
  return fg(pattern)
}

export function getSFCBlocks(
  descriptor: SFCDescriptor,
  version = 3
): SFCBlock[] {
  const { template, script, scriptSetup, styles, customBlocks } = descriptor
  const blocks: SFCBlock[] = [...styles, ...customBlocks]
  template && blocks.push(template as SFCBlock)
  script && blocks.push(script as SFCBlock)
  scriptSetup && blocks.push(scriptSetup as SFCBlock)
  blocks.sort((a, b) => {
    if (version === 3) {
      return a.loc.start.offset - b.loc.start.offset
    } else if (version === 2 && a.start != null && b.start != null) {
      return a.start - b.start
    } else {
      return 0
    }
  })
  return blocks
}

export function getSFCContentInfo(
  block: SFCBlock,
  filepath: string
): { content: string; lang: string | undefined; contentPath: string } {
  if (block.lang) {
    return { content: block.content, lang: block.lang, contentPath: filepath }
  } else if (block.attrs.lang && isString(block.attrs.lang)) {
    return {
      content: block.content,
      lang: block.attrs.lang,
      contentPath: filepath
    }
  } else if (block.src || (block.attrs.src && isString(block.attrs.src))) {
    // prettier-ignore
    const src = block.src
      ? block.src
      : block.attrs.src && isString(block.attrs.src)
        ? block.attrs.src
        : ''
    if (!src) {
      throw new Error(`src is empty`)
    }
    const parsed = path.parse(filepath)
    const target = path.resolve(parsed.dir, src)
    const parsedTarget = path.parse(target)
    return {
      content: readFileSync(target, 'utf8'),
      lang: parsedTarget.ext.split('.').pop(),
      contentPath: target
    }
  } else {
    return { content: block.content, lang: undefined, contentPath: filepath }
  }
}

export function getCustomBlockContenType(
  content: string
): CustomBlockContentType {
  const [isLoad, type] = loadJson(content)
  if (isLoad) {
    return type
  } else if (isLoadYaml(content)) {
    return 'yaml'
  } else {
    return 'unknwon'
  }
}

function isLoadYaml(content: string): boolean {
  try {
    parseYAML(content)
    return true
  } catch (e: unknown) {
    debug('yaml load error', (e as Error).message)
    return false
  }
}

function isLoadJson(content: string): boolean {
  try {
    JSON.parse(content)
    return true
  } catch (e: unknown) {
    debug('json load error', (e as Error).message)
    return false
  }
}

function isLoadJson5(content: string): boolean {
  try {
    parseJSON(content, { jsonSyntax: 'json5' })
    return true
  } catch (e: unknown) {
    debug('json5 load error', (e as Error).message)
    return false
  }
}

function loadJson(content: string): [boolean, 'json' | 'json5' | 'unknwon'] {
  try {
    if (isLoadJson(content)) {
      return [true, 'json']
    } else if (isLoadJson5(content)) {
      return [true, 'json5']
    } else {
      return [false, 'unknwon']
    }
  } catch (e: unknown) {
    debug('json load error', (e as Error).message)
    return [false, 'unknwon']
  }
}

const ESC: { [key in string]: string } = {
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  '&': '&amp;'
}

function escapeChar(a: string): string {
  return ESC[a] || a
}

export function escape(s: string): string {
  return s.replace(/[<>"&]/g, escapeChar)
}

const df = new diff.diff_match_patch()

export function hasDiff(newContent: string, oldContent: string): boolean {
  const diffs = df.diff_main(oldContent, newContent, true)
  if (diffs.length === 0) {
    return false
  }
  return !!diffs.find(
    d => d[0] === diff.DIFF_DELETE || d[0] === diff.DIFF_INSERT
  )
}

export function updateContents(
  original: string,
  contents: string[],
  offset: number,
  block: SFCBlock,
  vue: number
): [string[], number] {
  // prettier-ignore
  const end = vue === 3
    ? block.loc.end.offset
    : (vue === 2 && block.end != null)
      ? block.end
      : -1
  if (end === -1) {
    throw new Error('Not supported error')
  }
  const _contents = contents.concat(original.slice(offset, end))
  const _offset = end
  return [_contents, _offset]
}

export function getPosition(
  block: SFCBlock,
  vue: number,
  type: 'start' | 'end'
): [number, number] {
  if (vue === 3) {
    return [block.loc[type].offset, block.loc[type].column]
  } else if (vue === 2 && block[type] != null) {
    // @ts-ignore
    return [block[type], -1]
  } else {
    return [-1, -1]
  }
}

export function buildSFCBlockTag(meta: {
  type: string
  attrs: Record<string, any> // eslint-disable-line @typescript-eslint/no-explicit-any
}): string {
  let tag = `<${meta.type}`
  for (const [key, value] of Object.entries(meta.attrs)) {
    if (value === true) {
      tag += ` ${key}`
    } else {
      tag += ` ${key}="${escape(meta.attrs[key])}"`
    }
  }
  tag += '>'
  return tag
}

export async function getPrettierConfig(filepath: string) {
  const explorer = cosmiconfig('prettier')
  return await explorer.load(filepath)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const _rDefault = (r: any) => r.default || r

export async function getSFCParser(version: number): Promise<parseSFC | null> {
  let parser: parseSFC | null = null
  try {
    if (version === 3) {
      parser = (
        await import(
          path.resolve(__dirname, '../node_modules', '@vue/compiler-sfc')
        ).then(_rDefault)
      ).parse as parseSFC
    } else if (version === 2) {
      const { parseComponent } = await import(
        path.resolve(__dirname, '../node_modules', 'vue-template-compiler')
      ).then(_rDefault)
      parser = (source: string) => {
        const errors: SyntaxError[] = []
        const descriptor = parseComponent(source)
        return { descriptor, errors }
      }
    }
  } catch (e) {
    debug('getSFCParser error', e)
  }
  return parser
}
