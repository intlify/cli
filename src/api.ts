import path from 'path'
import { promises as fs } from 'fs'
import { debug as Debug } from 'debug'
import { generateJSON, generateYAML } from './generator/index'
import { globAsync } from './utils'

const SUPPORTED_FORMAT = ['.json', '.json5', '.yaml', '.yml']
const debug = Debug('@intlify/cli:api')

/**
 * Compile Error Codes
 *
 * @remarks
 * The error codes of {@link compile} function
 *
 * @public
 */
export const enum CompileErrorCodes {
  /**
   * Not supported format
   */
  NOT_SUPPORTED_FORMAT = 1,
  /**
   * Internal compile warning
   */
  INTERNAL_COMPILE_WARNING,
  /**
   * Internal compile error
   */
  INTERNAL_COMPILE_ERROR
}

/**
 * Compile Options
 *
 * @remarks
 * This optioins is used at {@link compile} function
 *
 * @public
 */
export interface CompileOptions {
  /**
   * Compile Error handler
   */
  onError?: (code: number, source: string, output: string, msg?: string) => void
  /**
   * Compile handler
   */
  onCompile?: (source: string, output: string) => void
}

/**
 * Compile i18n resources
 *
 * @param source - the i18n resource source path, you can use glob pattern
 * @param output - the compiled i18n resource output path
 * @param options - {@link CompileOptions}
 *
 * @remarks
 * This functoin is **asyncronous** function. If you want to get about error details, use the handler of {@link CompileOptions} and {@link CompileErrorCodes}
 *
 * @returns `true` when all i18n resource successfuly compile, not `false`
 *
 * @public
 */
export async function compile(
  source: string,
  output: string,
  options: CompileOptions = {}
): Promise<boolean> {
  let ret = true

  const targets = await globAsync(source)
  debug('compile: targets', targets)

  for (const target of targets) {
    const parsed = path.parse(target)
    debug('parsed', parsed)
    if (!parsed.ext) {
      continue
    }
    const filename = `${parsed.name}.js`
    const generatePath = path.resolve(output, filename)
    if (!SUPPORTED_FORMAT.includes(parsed.ext)) {
      options.onError &&
        options.onError(
          CompileErrorCodes.NOT_SUPPORTED_FORMAT,
          target,
          generatePath
        )
      ret = false
      continue
    }
    const source = await fs.readFile(target, { encoding: 'utf-8' })
    const generate = /\.json?5/.test(parsed.ext) ? generateJSON : generateYAML
    let occuredError = false
    const { code } = generate(source, {
      type: 'plain',
      filename: target,
      env: 'production',
      onError: (msg: string): void => {
        occuredError = true
        options.onError &&
          options.onError(
            CompileErrorCodes.INTERNAL_COMPILE_ERROR,
            target,
            generatePath,
            msg
          )
        ret = false
      },
      onWarn: (msg: string): void => {
        options.onError &&
          options.onError(
            CompileErrorCodes.INTERNAL_COMPILE_WARNING,
            target,
            generatePath,
            msg
          )
        ret = false
      }
    })
    if (!occuredError) {
      await writeGenerateCode(output, filename, code)
      options.onCompile && options.onCompile(target, generatePath)
    }
  }

  return ret
}

export async function writeGenerateCode(
  target: string,
  filename: string,
  code: string
): Promise<string> {
  await fs.mkdir(target, { recursive: true })
  const generatePath = path.resolve(target, filename)
  await fs.writeFile(generatePath, code)
  return generatePath
}
