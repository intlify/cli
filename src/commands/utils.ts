import path from 'pathe'
import { t } from '../i18n'
import { getSourceFiles } from '../utils'
import { RequireError } from './fail'

export function checkType(type?: string) {
  if (type !== 'custom-block') {
    throw new RequireError(
      t(`'--type' is not supported except for 'custom-block'`)
    )
  }
}

export function checkSource(argsLength: number, source?: string) {
  if (source == null && argsLength === 1) {
    throw new RequireError(
      t(
        `if you don't specify some files at the end of the command, the 'â€”-source' option is required`
      )
    )
  }
}

export async function getSFCFiles(
  source?: string,
  files?: (string | number)[]
) {
  return await getSourceFiles({ source, files }, file => {
    const parsed = path.parse(file as string)
    return parsed.ext === '.vue'
  })
}
