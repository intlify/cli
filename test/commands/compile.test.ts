import { beforeEach, afterEach, describe, it, vi, expect } from 'vitest'
import { prettier } from '../utils'
import { dirname, resolve } from 'pathe'
import { promises as fs, existsSync } from 'fs'
import { fileURLToPath } from 'url'
import yargs from 'yargs'
import { compile } from '../../src/commands'
import { initI18n } from '../../src/i18n'

const __dirname = dirname(fileURLToPath(import.meta.url))
const CWD = resolve(__dirname, '../../')
console.log('CWD', CWD)

let orgCwd
beforeEach(async function () {
  await initI18n()
  orgCwd = process.cwd
  process.cwd = vi.fn().mockImplementation(() => CWD)
})

afterEach(function () {
  process.cwd = orgCwd
})

describe('compile', function () {
  it('basic', async function () {
    // @ts-expect-error
    const cmd = yargs(compile())
    await cmd.parse(
      `compile --source ./test/fixtures/commands/compile-basic.json --output ./temp`
    )
    const expected = await fs.readFile(
      resolve(__dirname, '../__generated__/compile/commands/compile-basic.js'),
      'utf8'
    )
    console.log('expected', expected)
    console.log(
      'compiled ffile',
      existsSync(resolve(__dirname, '../../temp/compile-basic.js'))
    )
    const compiled = await fs.readFile(
      resolve(__dirname, '../../temp/compile-basic.js'),
      'utf8'
    )
    console.log('compiled', compiled)
    expect(await prettier(compiled)).toBe(await prettier(expected))
  })

  it('--mode development', async function () {
    // @ts-expect-error
    const cmd = yargs(compile())
    await cmd.parse(
      `compile --source ./test/fixtures/commands/compile-mode.json --output ./temp --mode development`
    )
    const expected = await fs.readFile(
      resolve(__dirname, '../__generated__/compile/commands/compile-mode.js'),
      'utf8'
    )
    const compiled = await fs.readFile(
      resolve(__dirname, '../../temp/compile-mode.js'),
      'utf8'
    )
    expect(await prettier(compiled)).toBe(await prettier(expected))
  })
})
