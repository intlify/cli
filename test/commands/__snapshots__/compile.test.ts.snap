// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile > ast format 1`] = `
{
  "backslash-backslash": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "\\\\\\\\",
      "t": 2,
    },
    "t": 0,
  },
  "backslash-single-quote": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "\\\\'",
      "t": 2,
    },
    "t": 0,
  },
  "complex": {
    "warnings": [
      {
        "b": {
          "i": [
            {
              "t": 3,
            },
          ],
          "s": "NOTE: This is warning",
          "t": 2,
        },
        "t": 0,
      },
      {
        "named-waring": {
          "b": {
            "i": [
              {
                "t": 3,
                "v": "this is ",
              },
              {
                "k": "type",
                "t": 4,
              },
              {
                "t": 3,
                "v": " warining",
              },
            ],
            "t": 2,
          },
          "t": 0,
        },
      },
    ],
  },
  "emoji": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "üò∫",
      "t": 2,
    },
    "t": 0,
  },
  "errors": [
    {
      "b": {
        "i": [
          {
            "t": 3,
          },
        ],
        "s": "ERROR1001",
        "t": 2,
      },
      "t": 0,
    },
    {
      "b": {
        "i": [
          {
            "t": 3,
          },
        ],
        "s": "ERROR1002",
        "t": 2,
      },
      "t": 0,
    },
  ],
  "hi": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "hi there!",
      "t": 2,
    },
    "t": 0,
  },
  "nested": {
    "hello": {
      "b": {
        "i": [
          {
            "t": 3,
          },
        ],
        "s": "hello world!",
        "t": 2,
      },
      "t": 0,
    },
    "list": {
      "b": {
        "i": [
          {
            "t": 3,
            "v": "hi, ",
          },
          {
            "i": 0,
            "t": 5,
          },
          {
            "t": 3,
            "v": " !",
          },
        ],
        "t": 2,
      },
      "t": 0,
    },
    "more": {
      "plural": {
        "b": {
          "c": [
            {
              "i": [
                {
                  "k": {
                    "t": 9,
                    "v": "no apples",
                  },
                  "m": {
                    "t": 8,
                    "v": "caml",
                  },
                  "t": 6,
                },
              ],
              "t": 2,
            },
            {
              "i": [
                {
                  "i": 0,
                  "t": 5,
                },
                {
                  "t": 3,
                  "v": " apple",
                },
              ],
              "t": 2,
            },
            {
              "i": [
                {
                  "k": "n",
                  "t": 4,
                },
                {
                  "t": 3,
                  "v": " apples",
                },
              ],
              "t": 2,
            },
          ],
          "t": 1,
        },
        "t": 0,
      },
    },
  },
  "single-quote": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "I don't know!",
      "t": 2,
    },
    "t": 0,
  },
  "unicode": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "A",
      "t": 2,
    },
    "t": 0,
  },
  "unicode-escape": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "\\\\u0041",
      "t": 2,
    },
    "t": 0,
  },
  "„Åì„Çì„Å´„Å°„ÅØ": {
    "b": {
      "i": [
        {
          "t": 3,
        },
      ],
      "s": "„Åì„Çì„Å´„Å°„ÅØÔºÅ",
      "t": 2,
    },
    "t": 0,
  },
}
`;
